### ISSUE TITLE
Hundreds wrong detection on yolov8

### ISSUE THREAD

**Issue by Son-kun** [deepjavalibrary/djl#2968]
Using code writed for yolov5, after some adjustment for yolov8, when I try to use torchscript model generated I got ~200 detection with very high score (10000+), and they are mostly focused in left top corner. My model probably is created correctly using yolov8s/yolov8n model (Ultralytics tool) and `yolo predict` work great on that model. Then I exported to torchscript format (`yolo export model=... format=torchscript` like v5 version).

![1-Figure1-1](https://github.com/deepjavalibrary/djl/assets/5764004/43c529d9-a7ee-4921-a40e-c55639635cbd)

I updated djl to 0.26 and changed translator to `YoloV8Translator`.
My yolo usage looks like:
```
        Image picture = ImageFactory.getInstance().fromFile(image.toPath());
        
        Pipeline pipeline = new Pipeline();
        pipeline.add(new Resize(640));
        pipeline.add(new ToTensor());

        List<String> synset = new ArrayList<>(4);
        synset.add(DETECTION_LABEL_0);
        synset.add(DETECTION_LABEL_1);
        synset.add(DETECTION_LABEL_2);
        synset.add(DETECTION_LABEL_3);

        YoloV8Translator translator = YoloV8Translator
                .builder()
                .setPipeline(pipeline)
                .optSynset(synset)
                .optThreshold(0.7f)
                .build();

        Criteria criteria = Criteria.builder()
                .setTypes(Image.class, DetectedObjects.class)
                .optDevice(Device.cpu())
                .optModelPath(Paths.get(modelLocation))
                .optTranslator(translator)
                .optEngine("PyTorch")
                .build();

        ZooModel<Image, DetectedObjects> model = criteria.loadModel();
        Predictor<Image, DetectedObjects> predictor = model.newPredictor();
        predictor.predict(picture);
```

Same code with YoloV5Translator, djl 0.19.0 and model built on yolov5 works fine.
What I forgot, or what I do wrong? Can you help me migrate to v8?

**Comment by frankfliu** [comment]
Can you try to use `YoloV8TranslatorFactory`, it will set a few default values for the translator:

```
        Criteria<Image, DetectedObjects> criteria =
                Criteria.builder()
                        .setTypes(Image.class, DetectedObjects.class)
                        .optModelPath(Paths.get(modelLocation))
                        .optArgument("width", 640)
                        .optArgument("height", 640)
                        .optArgument("resize", true)
                        .optArgument("toTensor", true)
                        .optArgument("applyRatio", true)
                        .optArgument("threshold", 0.7f)
                        .optArgument("synset", "LABEL_1,LABEL_2,LABEL_3...")
                        // for performance optimization maxBox parameter can reduce number of
                        // considered boxes from 8400
                        .optArgument("maxBox", 1000)
                        .optTranslatorFactory(new YoloV8TranslatorFactory())
                        .build();

```

**Comment by Son-kun** [comment]
I had to add engine (`PyTorch`) for it to run at all.

When I use maxBox, 1000 I dont get any detection, when I increase it to 4k or leave default (8400) then I got exception:
```
java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 4
```
Error occurs in YoloV8Translator class in method processFromBoxOutput when it want to create new IntermediateResult with accessing to list of classes (my labels) - line 91.
I'm not sure why it want to get higher number than elements in classes (labels) declared.

https://github.com/deepjavalibrary/djl/blob/5b4cf0e9be0e9c1415b27ca3542ef76208af7c56/api/src/main/java/ai/djl/modality/cv/translator/YoloV8Translator.java#L70-L92

**Comment by zachgk** [comment]
It could be possible that your model is using a different output format than the one expected by the `Yolov8Translator`. What is the shape of the output of your model?

**Comment by Son-kun** [comment]
I never set any shape. For training I used rectangle on images.
When I run predict in ultralytics tool, as output I also get images with rectangle.
So how I can tell what shape is set in my model, and hot to change it? I can't find anything about that on yolov8 github page...

**Comment by zachgk** [comment]
I mean the `NDArray.shape()` for the arrays in the `NDList` output of your model (or `NDList.getShapes()`). What are those shape?

**Comment by SidneyLann** [comment]
It should be maxIndex = c -4 ; but not maxIndex = c;  in YoloV8Translator.java

**Comment by Son-kun** [comment]
So, my shape is list of:
[0] = 8400
[1] = 8
I didn't change anythink from default settings like in yolov5. Should I change any param in training process?

About change maxIndex (rebuild api module localy) from `c` to `c - 4`, It's working good :)
@SidneyLann - it's a generic bug in Yolov8Translator, or it's depends of some corner-case?

**Comment by SidneyLann** [comment]
It's a generic bug

**Comment by zachgk** [comment]
@Son-kun @SidneyLann Do either of you want to raise a PR to fix this?

**Comment by frankfliu** [comment]
I don't think `maxIndex = c - 4` is correct. It's not a bug in the code.

Your output shape contains 8 classes, but you only provided 4 labels.  You should do:

```
.optArgument("synset", "LABEL_1,LABEL_2,LABEL_3,LABEL_4,LABEL_5,LABEL_6,LABEL7")
```

**Comment by frankfliu** [comment]
If you trained with 4 classes, the output Tensor shape should be (8400, 4). Can you check your training script?

**Comment by SidneyLann** [comment]
The amending is not correct.

**Comment by urieli** [comment]
@Son-kun I'm having similar issues with Yolov8. The detection works perfectly in Python using the Ultralytics library. In Java with DJL, not only are the classes offset by 4, but I'm not getting the same rectangles detected. I'm using the OnnxRuntime engine, as I wasn't able to get the native PyTorch engine working at all (either with a .pt file or with a torchscript archive). If you manage to get your Yolov8 model working correctly in DJL, please let me know.

**Comment by hohonuuli** [comment]
I'm hitting this same issue. The code I'm using to build the Criteria is:

```java
    public static Criteria<Image, DetectedObjects> buildCriteria(Path modelPath, Path namesPath, int resolution, int yoloVersion) {

        var names = NamesUtil.load(namesPath);

        Pipeline pipeline = new Pipeline();
        pipeline.add(new Resize(resolution)); 
        pipeline.add(new ToTensor());

        Translator<Image, DetectedObjects> translator;
        if (yoloVersion == 8) {
            translator = YoloV8Translator.builder()
                    .setPipeline(pipeline)
                    .optSynset(names)
                    .build();
        }
        else {
            translator = YoloV5Translator.builder()
                    .setPipeline(pipeline)
                    .optSynset(names)
                    .build();
        }


        return Criteria.builder()
                .optApplication(Application.CV.OBJECT_DETECTION)
                .setTypes(Image.class, DetectedObjects.class)
                .optModelPath(modelPath)
                .optTranslator(translator)
                .optEngine("PyTorch")
                .build();
    }

```

Here's some example results using yolov5 and yolov8 models converted from .pt files to torchscript:

### Yolov5

![112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b-detections-5](https://github.com/deepjavalibrary/djl/assets/282941/3415febc-57dc-4810-8636-006ebc22151d)

### Yolov8
![112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b-detections-8](https://github.com/deepjavalibrary/djl/assets/282941/ea3775af-fff9-4626-95ad-04f8e512f147)

**Comment by frankfliu** [comment]
@hohonuuli 

Can you provide reproduce project I can test with (include model and test image)?

**Comment by hohonuuli** [comment]
@frankfliu I can. I'm heading out of the office now, but will add a sample for you tomorrow.

**Comment by hohonuuli** [comment]
@frankfliu You can replicate this using:

```sh
# requires idk 17+ and maven
git clone git@github.com:mbari-org/pythia.git
cd pythia

# Runs yolov5 model. Output will be in ./target/112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b-detections-5.png
mvn test -Dtest=Yolov8ServiceTest

# Runs yolov8 model. Output will be in ./target/112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b-detections-8.png
mvn test -Dtest=Yolov8ServiceTest
```

The code of interest is mostly in `YoloService.java`. The models are in `src/test/resources/models/` as torchscript. The test image is at `src/test/resources/images/112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b.jpg`  I've verified the `mbari401k_yolov8.torchscript` model works using ultralytics with:

```sh
yolo predict model=mbari401k_yolov8.torchscript source=112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b.jpg imgsz=640
```

The output from ultralytics is:

![112512--1c22e634-2378-4b66-86e4-e00d3ca3ad5b](https://github.com/deepjavalibrary/djl/assets/282941/22b53850-d6fa-4883-acff-fe9fcf7a7520)

**Comment by frankfliu** [comment]
OK, this is a silly bug in Yolo8TranslatorBuilder. I created a PR to fix this issue.

There are two ways to workaround this issue:
1. use YoloV8TranslatorFatory, and this is recommended way to create Criteria: see: https://github.com/deepjavalibrary/djl/blob/master/examples/src/main/java/ai/djl/examples/inference/Yolov8Detection.java#L56-L65
2. Use `new YoloV8Translator.Builder()` instead of `YoloV8Translator.builder()`
```
translator = new YoloV8Translator.builder()
```

**Comment by hohonuuli** [comment]
@frankfliu Thank you!

**Comment by hohonuuli** [comment]
@frankfliu One more questions ...

When I try to use the recommended way to create a Critera as described [here](https://github.com/deepjavalibrary/djl/blob/master/examples/src/main/java/ai/djl/examples/inference/Yolov8Detection.java#L56-L65), it expects a file name `synset.txt` in the same location as the model, How can I pass the list the makes up the synset into the Criteria builder? I was using the code below, but it's not clear how to pass a list of names using the recommended approach:

```java
List<String> names = // load a list of names myself
translator = new YoloV8Translator.Builder()
                .setPipeline(pipeline)
                .optSynset(names)
                .build();
```

**Comment by frankfliu** [comment]
```
.optArgument("synset", String.joint(",", names)
```

or 
```
.optArgument("synsetUrl", "file:/")
```
